### ISSUE TITLE
Running tests uses 50.1GB of disk space on Ubuntu

### ISSUE THREAD

**Issue by samuelcolvin** [apache/datafusion#11105]
### Describe the bug

I just cloned datafusion and tried `cargo t` on my ubuntu desktop, to check things were working properly.

It crashed.

I restarted, and it seems datafusion is using 50.1GB to just run tests.

The `examples` directory seems to be the biggest culpret.

![image](https://github.com/apache/datafusion/assets/4039449/abe755ab-44a2-4b41-bbfa-c98ce28bd8b2)

### To Reproduce

Just clone and run `cargo t` on Ubuntu (no idea if this is limited to linux).

### Expected behavior

Humm, I guess in principle this isn't a show stopper, but seems somewhat unfortunate.

If there's an easy/low impact way to reduce disk usage, it might be useful.

**Comment by jayzhan211** [comment]
I got 28GB after `cargo clean` and `cargo t`

<img width="814" alt="Screenshot 2024-06-25 at 5 23 02 PM" src="https://github.com/apache/datafusion/assets/15869383/3e7df559-af17-4e4a-acae-ca5872ceb56f">

**Comment by jcsherin** [comment]
On Ubuntu,

```sh
$ lsb_release -rc
Release:	22.04
Codename:	jammy
```
After `cargo clean` and running `cargo t`:

```sh
$ du -h -d2 target
4.0K	target/tmp
6.1G	target/debug/incremental
12M	target/debug/.fingerprint
337M	target/debug/build
16G	target/debug/deps
28G	target/debug/examples
50G	target/debug
50G	target
```

**Comment by samuelcolvin** [comment]
![image](https://github.com/apache/datafusion/assets/4039449/8faabb73-1d73-4566-a8ff-680c85fc139f)

Update, not using 103GB.

Seeing this nice error:

```
...
   Compiling datafusion v39.0.0 (/home/samuel/code/datafusion/datafusion/core)
error: failed to write to `/home/samuel/code/datafusion/target/debug/deps/rmetalvEs2G/lib.rmeta`: No space left on device (os error 28)

error: could not compile `datafusion` (lib) due to 1 previous error
```

**Comment by jayzhan211** [comment]
The more you run, the more artifacts in /target is. `cargo clean` is all you need.

**Comment by alamb** [comment]
Yeah, I think this obscene use of temp space is an artifact of rust (rather than an artifact of datafusion specifically)

**Comment by alamb** [comment]
It would also help if we consolidated some of the example binaries: https://github.com/apache/datafusion/issues/11172

**Comment by alamb** [comment]
I filed a few tickets to start consolidating the examples which i think will help: 
- https://github.com/apache/datafusion/issues/13842
- https://github.com/apache/datafusion/issues/13844
- https://github.com/apache/datafusion/issues/13843
